### PR TITLE
Introduce removing inactive core command

### LINKED ISSUE #660: Force parameter in 'casadm -R' command removes active dirty core
## Description
Force parameter in 'casadm -R' command removes active dirty core, but help says that it's a command for removing inactive core devices.

## Possible Fix
Change help message or create two separate methods (parameters) for removing dirty active and inactive core devices

## Your Environment
* OpenCAS version (commit hash or tag): ff00dab04cf8e0f302ed8969ca9ce2050641c8a0

### PR DESCRIPTION
**Pull request by Ostrokrzew**
Split the `casadm --remove-core` command into two separate commands for different core states.

Related: #660

### CODE REVIEW

File: casadm/cas_lib.c [hunk 1]
```diff
@@ -1905,11 +1905,22 @@ int remove_core(unsigned int cache_id, unsigned int core_id,
 			"Removing core", cache_id, core_id) < 0) {
 		close(fd);
 		if (OCF_ERR_FLUSHING_INTERRUPTED == cmd.ext_err_code) {
-			cas_printf(LOG_ERR, "You have interrupted removal of core. CAS continues to operate normally.\n");
+			if (detach) {
+				cas_printf(LOG_ERR, "You have interrupted "
+						"detachment of core. CAS "
```
Comment (arutk) [inline casadm/cas_lib.c:1910]: I think "detachment" has a different meaning, "detaching" would work fine

File: casadm/cas_lib.c [hunk 2]
```diff
@@ -1905,11 +1905,22 @@ int remove_core(unsigned int cache_id, unsigned int core_id,
...
+				cas_printf(LOG_ERR, "You have interrupted "
+						"removal of core. CAS "
+						"continues to operate "
+						"normally.\n");
+			}
 			return INTERRUPTED;
 		} else {
-			cas_printf(LOG_ERR, "Error while removing core device %d from cache instance %d\n",
-				   core_id, cache_id);
+			cas_printf(LOG_ERR, "Error while removing core device "
```
Comment (arutk) [inline casadm/cas_lib.c:1921]: how about making similar distinction between detaching and removal here?

File: casadm/cas_main.c [hunk 3]
```diff
@@ -1097,7 +1097,7 @@ int handle_add()
 static cli_option remove_options[] = {
 	{'i', "cache-id", CACHE_ID_DESC, 1, "ID", CLI_OPTION_REQUIRED},
 	{'j', "core-id", CORE_ID_DESC, 1, "ID", CLI_OPTION_REQUIRED},
-	{'f', "force", "Force remove inactive core"},
+	{'f', "force", "Force remove active core without data flush"},
```
Comment (arutk) [inline casadm/cas_main.c:1117]: *removal?

File: casadm/cas_lib.h [hunk 4]
```diff
@@ -215,6 +215,15 @@ int get_core_info(int fd, int cache_id, int core_id, struct kcas_core_info *info
...
 		bool detach, bool force_no_flush);
 
+/**
+ * @brief remove inactive core device from a cache
+ *
+ * @param cache_id cache from which inactive core is being removed
+ * @param cache_id inactive core which is being removed
+ * @return 0 upon successful core addition, 1 upon failure
+ */
+int remove_inactive_core(unsigned int cache_id, unsigned int core_id);
```
Comment (arutk) [inline casadm/cas_lib.h:225]: please state command name in the commit message

File: casadm/cas_lib.h [hunk 5]
```diff
@@ -215,6 +215,15 @@ int get_core_info(int fd, int cache_id, int core_id, struct kcas_core_info *info
 int remove_core(unsigned int cache_id, unsigned int core_id,
 		bool detach, bool force_no_flush);
 
+/**
+ * @brief remove inactive core device from a cache
+ *
+ * @param cache_id cache from which inactive core is being removed
+ * @param cache_id inactive core which is being removed
+ * @return 0 upon successful core addition, 1 upon failure
```
Comment (arutk) [inline casadm/cas_lib.h:223]: addition->removal

File: modules/cas_cache/layer_cache_management.c [hunk 6]
```diff
@@ -1422,6 +1417,15 @@ int cache_mngt_remove_core_from_cache(struct kcas_remove_core *cmd)
 	if (result)
 		return result;
 
+	result = _cache_mngt_remove_core_is_active(cache, cmd->core_id);
+	if (result) {
+		printk(KERN_WARNING OCF_PREFIX_SHORT
+				"Core is already inactive. To manage the "
```
Comment (arutk) [inline modules/cas_cache/layer_cache_management.c:1423]: I would remove the word "already" here, this suggests that this is the target state of a core.

File: modules/cas_cache/layer_cache_management.c [hunk 7]
```diff
@@ -1422,6 +1417,15 @@ int cache_mngt_remove_core_from_cache(struct kcas_remove_core *cmd)
...
+				"inactive core use '--remove-inactive' "
+				"command.\n");
+		goto put;
```
Comment (arutk) [inline modules/cas_cache/layer_cache_management.c:1426]: it would be more user friendly to print this message in casadm - this would require adding new return code.

File: modules/cas_cache/layer_cache_management.c (hunk 6 above)
Comment (arutk) [inline modules/cas_cache/layer_cache_management.c:1423]: This is not necessarily true. Any arror in _cache_mngt_remove_core_is_active would trigger this, not only inactive core message.

File: modules/cas_cache/layer_cache_management.c [hunk 8]
```diff
@@ -1411,6 +1385,27 @@ static void _cache_mngt_remove_core_complete(void *priv, int error)
 	complete(&context->cmpl);
 }
 
+static int _cache_mngt_remove_core_is_active(ocf_cache_t cache,
```
Comment (arutk) [inline modules/cas_cache/layer_cache_management.c:1388]: Given the implementation, it would make sense to rename this function to cache_mngt_check_core_active

File: modules/cas_cache/service_ui_ioctl.c [hunk 9]
```diff
@@ -132,6 +132,16 @@ long cas_service_ioctl_ctrl(struct file *filp, unsigned int cmd,
 		RETURN_CMD_RESULT(cmd_info, arg, retval);
 	}
 
+	case KCAS_IOCTL_REMOVE_INACTIVE: {
```
Comment (arutk) [inline modules/cas_cache/service_ui_ioctl.c:135]: please state the IOCTL name instead of "new" in the commit message

Comment (arutk) [reply modules/cas_cache/service_ui_ioctl.c:135]: also, our adding just the struct (despite what the title suggests)

**Review by arutk — CHANGES_REQUESTED** [review]

File: modules/cas_cache/service_ui_ioctl.c [hunk 10]
```diff
@@ -132,6 +132,16 @@ long cas_service_ioctl_ctrl(struct file *filp, unsigned int cmd,
...
+		struct kcas_remove_inactive *cmd_info;
+
+		GET_CMD_INFO(cmd_info, arg);
+
+		retval = cache_mngt_remove_inactive_core_from_cache(cmd_info);
```
Comment (arutk) [inline modules/cas_cache/service_ui_ioctl.c:140]: "from_cache" suffix seems redundant

File: casadm/extended_err_msg.c [hunk 11]
```diff
@@ -158,6 +158,10 @@ struct {
 		OCF_ERR_INVAL_CACHE_DEV,
 		"Device does not meet requirements."
 	},
+	{
```
Comment (arutk) [inline casadm/extended_err_msg.c:161]: Like mentioned elsewhere: this should probaby be KCAS error

File: modules/cas_cache/layer_cache_management.c [hunk 12]
```diff
@@ -1327,7 +1327,6 @@ static int _cache_mngt_remove_core_flush(ocf_cache_t cache,
 {
 	int result = 0;
```
Comment (arutk) [inline modules/cas_cache/layer_cache_management.c:1328]: This commit does more than just "refactoring" - more of a redesign. or maybe:
remove inactive core handling from "remove core"

File: modules/cas_cache/layer_cache_management.c [hunk 13]
```diff
@@ -1346,59 +1345,34 @@ static int _cache_mngt_remove_core_flush(ocf_cache_t cache,
...
+	if (result)
+		return result;
 
-	if (core_active) {
-		result = block_dev_destroy_exported_object(core);
-		if (result)
-			return result;
-	}
+	if (cmd->force_no_flush)
+		return -KCAS_ERR_REMOVED_DIRTY;
```
Comment (arutk) [inline modules/cas_cache/layer_cache_management.c:1373]: This is an important change (IOCTL handler returning KCAS_ERR_REMOVED_DIRTY for "casadm -R -f"). I would make a separate commit out of this.

File: modules/cas_cache/layer_cache_management.c (hunk 6 above)
Comment (arutk) [inline modules/cas_cache/layer_cache_management.c:1423]: This is "time of check vs time of use" kind of error. After you check whether core is active, its state might change. You should not release the management lock between checking core active state and removing the core. This means you should do both under one management write lock

**Review by arutk — CHANGES_REQUESTED** [review]

File: modules/cas_cache/layer_cache_management.c (hunk 12 above)
Comment (Ostrokrzew) [reply modules/cas_cache/layer_cache_management.c]: I rolled back most of the changes. Existing code was better.

File: modules/cas_cache/layer_cache_management.c (hunk 6 above)
Comment (Ostrokrzew) [reply modules/cas_cache/layer_cache_management.c]: That's a very important advice. Thank you.

**Comment by robertbaldyga** [comment]
I suggest squashing commits titled "part x" into one.

File: casadm/cas_lib.c [hunk 14]
```diff
@@ -1919,6 +1946,43 @@ int remove_core(unsigned int cache_id, unsigned int core_id,
...
 }
 
+int remove_inactive_core(unsigned int cache_id, unsigned int core_id)
+{
+	int fd = 0;
+	struct kcas_remove_inactive cmd;
+
+	/* don't even attempt ioctl if filesystem is mounted */
+	if (SUCCESS != check_if_mounted(cache_id, core_id)) {
+		return FAILURE;
```
Comment (arutk) [inline casadm/cas_lib.c:1956]: please add some error message here

**Review by arutk — CHANGES_REQUESTED** [review]
looks good to me, just a minor comment

File: casadm/cas_lib.c (hunk 14 above)
Comment (Ostrokrzew) [reply casadm/cas_lib.c:1956]: Error message is inside `check_if_mounted`.

**Comment by Ostrokrzew** [comment]
retest this please

**Comment by Ostrokrzew** [comment]
retest this please

**Review by arutk — DISMISSED** [review]

File: casadm/cas_lib.c [hunk 15]
```diff
@@ -1904,12 +1904,39 @@ int remove_core(unsigned int cache_id, unsigned int core_id,
...
+				cas_printf(LOG_ERR, "You have interrupted "
+						"detaching of core. CAS "
+						"continues to operate "
+						"normally.\n");
+			} else {
+				cas_printf(LOG_ERR, "You have interrupted "
+						"removal of core. CAS "
+						"continues to operate "
+						"normally.\n");
+			}
```
Comment (imjfckm) [inline casadm/cas_lib.c:1918]: You could deduplicate this by writing `cas_printf(LOG_ERR, "You have... %s ... \n", "detaching" ? detach : "removal")`

Comment (Ostrokrzew) [reply casadm/cas_lib.c:1918]: Thanks! It would be more aesthetic.

Comment (Ostrokrzew) [reply casadm/cas_lib.c:1918]: Thanks! It would be more aesthetic.

**Comment by Ostrokrzew** [comment]
retest this please

**Comment by Ostrokrzew** [comment]
retest this please

**Review by arutk — DISMISSED** [review]

**Review by jfckm — APPROVED** [review]

**Review by jfckm — APPROVED** [review]
